File: app/src/main/java/com/example/admin/indentifycode/DubCountDownTimer.java
package com.example.admin.indentifycode;

import android.os.CountDownTimer;
import android.util.Log;

/**
 * Created by admin on 2017/6/23.
 * 显示倒计时
 */

public class DubCountDownTimer {

    private CountDownTimer mCountDown;

    public DubCountDownTimer(CountDownTimerListener countDownTimerListener) {
        this.countDownTimerListener = countDownTimerListener;
    }

    public void startCountDown(long millis, long interval) {
        if (null == mCountDown) {
            mCountDown = new CountDownTimer(millis, interval) {
                @Override
                public void onTick(long millisUntilFinished) {
                    if (null != countDownTimerListener)
                        countDownTimerListener.tickTimer(millisUntilFinished);
                }

                @Override
                public void onFinish() {
                    if (null != countDownTimerListener) countDownTimerListener.finishTimer();
                }
            }.start();
            if (null != countDownTimerListener) countDownTimerListener.startTimer();
        } else {
            Log.i("         ", "倒计时中....");
        }
    }

    public void cancleCountDown() {
        if (mCountDown != null) {
            mCountDown.cancel();
            mCountDown = null;
        }
        if (countDownTimerListener != null) countDownTimerListener.cancelTimer();
    }

    private CountDownTimerListener countDownTimerListener; //这时创建了一个接回调，分别就情况而处理

    interface CountDownTimerListener {
        void startTimer();

        void finishTimer();

        void cancelTimer();

        void tickTimer(long l);
    }
}

File: appkotlin2/src/main/java/com/example/appkotlin/DubCountDownTimer.kt
package com.example.appkotlin

import android.os.CountDownTimer
import android.util.Log

/**
 * Created by admin on 2017/6/23.
 */

class DubCountDownTimer constructor(private var countDownListener: CountDownListener?) {

    private var mCounntDownTimer: CountDownTimer? = null
    fun startTimer(millisInFuture: Long, countDownInterval: Long) {
        if (null == mCounntDownTimer ) {
            countDownListener?.startTimer()
            mCounntDownTimer = object : CountDownTimer(millisInFuture, countDownInterval) {
                override fun onFinish() {
                    countDownListener?.finishTimer()
                    Log.i(Tags.toString(), "订时结束")
                }

                override fun onTick(millisUntilFinished: Long) {
                    countDownListener?.tickTimer(millisUntilFinished)
                    Log.i(Tags.toString(), "订时开始")
                }
            }.start()
        } else {
            countDownListener?.finishTimer()
            Log.i(Tags.toString(), "倒计时正在进行")
        }
    }

    fun ancleTimers() {
        if (mCounntDownTimer != null) {
            mCounntDownTimer!!.cancel()
            mCounntDownTimer = null
        }
        countDownListener?.cancleTimer()
    }

    interface CountDownListener {
        fun startTimer()
        fun cancleTimer()
        fun finishTimer()
        fun tickTimer(millisUntilFinished: Long)
    }

    companion object {
        private var Tags = DubCountDownTimer::class.java
    }
}

